### PR TITLE
Show Project Resource Quota

### LINKED ISSUE #915: Show Project Resource Quota
**What would you like to be added**:
With https://github.com/gardener/gardener/pull/2627 it is possible to set-up project related resource quotas to constraint the amount of objects e.g., `Shoots`, `Secrets`, `SecretBindings`, stakeholders can create in the Garden cluster. As a usability feature it'd be helpful to show the current quota status of the project to project members.

**Why is this needed**:
End users can easily see the current status of their quotas.

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
This PR adds support to show project resource quotas

Screen Shot 2022-08-29 at 13 40 33

**Which issue(s) this PR fixes**:
Fixes #915

**Special notes for your reviewer**:

**Release note**:

```feature user
Show project resource quotas
```

### CODE REVIEW

**Comment by petersutter** [comment]
nice with the card on the administration page! However I don't think the main toolbar is a good fit for displaying the quota as there is only place for one quota information. Maybe it sufficient to see the quota just on the administration page or an error if we know that the quota will not be sufficient for creating a shoot, infrastructure secret (secret + secretBinding), or service account.

File: frontend/src/components/MainToolbar.vue [hunk 1]
```diff
@@ -189,7 +189,8 @@ export default {
       'username',
       'displayName',
       'avatarUrl',
-      'isAdmin'
+      'isAdmin',
+      'projectFromProjectList'
```
Comment (holgerkoser) [inline frontend/src/components/MainToolbar.vue:193]: remove

File: backend/jest.setup.js [hunk 2]
```diff
@@ -109,15 +109,16 @@ jest.mock('./lib/config/gardener', () => {
...
 
 jest.mock('./lib/cache', () => {
-  const { find } = require('lodash')
+  const { find, split } = require('lodash')
   const fixtures = require('./__fixtures__')
   const originalCache = jest.requireActual('./lib/cache')
   const createTicketCache = jest.requireActual('./lib/cache/tickets')
   const { cache } = originalCache
-  const keys = ['cloudprofiles', 'seeds', 'quotas', 'projects', 'controllerregistrations']
+  const keys = ['core.gardener.cloud/cloudprofiles', 'core.gardener.cloud/seeds', 'core.gardener.cloud/quotas', 'core.gardener.cloud/projects', 'core.gardener.cloud/controllerregistrations', 'core/resourcequotas']
```
Comment (holgerkoser) [inline backend/jest.setup.js:117]:
```suggestion
  const keys = [
    'core.gardener.cloud/cloudprofiles', 
    'core.gardener.cloud/seeds', 
    'core.gardener.cloud/quotas', 
    'core.gardener.cloud/projects', 
    'core.gardener.cloud/controllerregistrations', 
    'core/resourcequotas'
  ]
```

File: frontend/src/utils/index.js [hunk 3]
```diff
@@ -317,6 +320,57 @@ export function getProjectDetails (project) {
   }
 }
 
+export function getProjectQuotaStatus (project) {
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/utils/index.js:323]: I have added it to the utils as the other related methods are also there and it makes it easier to test the function if it is separated from the component

File: frontend/src/views/Administration.vue [hunk 4]
```diff
@@ -547,6 +611,18 @@ export default {
       this.detailedMessage = undefined
       this.edit = false
     }
+  },
+  created () {
+    // watch the params of the route to fetch the data again
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/views/Administration.vue]: added link to docu

[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/views/Administration.vue]: added link to docu

File: frontend/src/store/modules/projects.js [hunk 5]
```diff
@@ -49,6 +54,21 @@ const actions = {
     const namespace = metadata.namespace
     await deleteProject({ namespace })
     // do not remove project from store as it will stay in termininating phase for a while
+  },
+  async getProjectQuota ({ commit, rootState }, { namespace, name }) {
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/store/modules/projects.js]: done

File: frontend/src/store/modules/projectQuota.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
+//
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
```
Comment (petersutter) [inline frontend/src/store/modules/projectQuota.js:2]:
```suggestion
// SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Gardener contributors
```

File: backend/jest.setup.js [hunk 7]
```diff
@@ -113,15 +113,23 @@ jest.mock('./lib/config/gardener', () => {
...
+  const keys = [
+    'core.gardener.cloud/cloudprofiles',
+    'core.gardener.cloud/seeds',
+    'core.gardener.cloud/quotas',
+    'core.gardener.cloud/projects',
+    'core.gardener.cloud/controllerregistrations',
+    'core/resourcequotas'
+  ]
   for (const key of keys) {
+    const [, fixtureName] = split(key, '/')
```
Comment (holgerkoser) [inline backend/jest.setup.js:130]:
```suggestion
    const [, resourceName] = key.split('/')
```

File: backend/lib/cache/index.js [hunk 8]
```diff
@@ -102,13 +106,20 @@ module.exports = {
...
+  },
   findProjectByNamespace (namespace) {
-    const project = cache.get('projects').find(['spec.namespace', namespace])
+    const project = cache.get('core.gardener.cloud/projects').find(['spec.namespace', namespace])
     if (!project) {
       throw new NotFound(`Namespace '${namespace}' is not related to a gardener project`)
     }
     return project
   },
+  findResourceQuotasByNamespace (namespace) {
```
Comment (holgerkoser) [inline backend/lib/cache/index.js:119]: The name of this method it wrong. But however I think we should remove it completely do the following where it is use:
```js
const resourceQuotas = _.filter(cache.getResourceQuotas(), ['metadata.namespace', namespace])
```

File: backend/lib/cache/index.js [hunk 9]
```diff
@@ -21,27 +21,31 @@ class Cache extends Map {
...
     return this.get('shoots').list()
   }
 
   getControllerRegistrations () {
-    return this.get('controllerregistrations').list()
+    return this.get('core.gardener.cloud/controllerregistrations').list()
+  }
+
+  getResourceQuotas () {
+    return this.get('core/resourcequotas').list()
```
Comment (holgerkoser) [inline backend/lib/cache/index.js:48]: The access methods on the cache do not have an apiGroup prefix. We have no duplicate names. Why do we need the apiGroup internally. Shouldn't we remove it? What do you think?

File: backend/lib/hooks.js [hunk 10]
```diff
@@ -60,7 +62,12 @@ class LifecycleHooks {
   static createInformers (client) {
     const informers = {}
     for (const key of this.resources) {
-      const observable = client['core.gardener.cloud'][key]
+      const [apiGroup, name] = _.split(key, '/')
+      if (!apiGroup || !name) {
+        assert.fail('Invalid resource key. Need to have format apiGroup/resourceName.')
+      }
```
Comment (holgerkoser) [inline backend/lib/hooks.js:68]: And may be do something like this?
```js
      let name, apiGroup
      if (['cloudprofiles', 'quotas', 'seeds', 'shoots', 'projects', 'controllerregistrations']).includes(key) {
         name = key
         apiGroup = 'core.gardener.cloud'
      } else if (['resourcequotas'].includes(key)) {
         name = key
         apiGroup = 'core'
      } else {
         [apiGroup, name] = key.split('/')
      }
      if (!apiGroup || !name) {
        assert.fail('Invalid resource key. Need to have format apiGroup/resourceName.')
      }
```

File: backend/lib/routes/index.js [hunk 11]
```diff
@@ -21,7 +21,8 @@ module.exports = {
   '/namespaces/:namespace/shoots': require('./shoots'),
   '/namespaces/:namespace/tickets': require('./tickets'),
   '/namespaces/:namespace/cloud-provider-secrets': require('./cloudProviderSecrets'),
-  '/namespaces/:namespace/members': require('./members')
+  '/namespaces/:namespace/members': require('./members'),
+  '/namespaces/:namespace/resource-quotas': require('./resourceQuotas')
```
Comment (holgerkoser) [inline backend/lib/routes/index.js:25]: In the k8s api the there is no `-`. I know we have them also in `cloud-provider-secrets` but not in `cloudprofiles`.

File: backend/lib/services/index.js [hunk 12]
```diff
@@ -15,5 +15,6 @@ module.exports = {
   authentication: require('./authentication'),
   tickets: require('./tickets'),
   terminals: require('./terminals'),
-  controllerregistrations: require('./controllerregistrations')
+  controllerregistrations: require('./controllerregistrations'),
+  resourceQuotas: require('./resourceQuotas')
```
Comment (holgerkoser) [inline backend/lib/services/index.js:19]: I think it should be all lowercase?

File: backend/lib/services/resourceQuotas.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  })
+}
+
+exports.list = async function ({ user, namespace }) {
+  const allowed = await authorization.canListResourceQuotas(user, namespace)
+  if (!allowed) {
+    throw new Forbidden(`You are not allowed to list resource quotas in namespace ${namespace}`)
+  }
+
+  const resourceQuotas = findResourceQuotasByNamespace(namespace)
```
Comment (holgerkoser) [inline backend/lib/services/resourceQuotas.js:30]: inline the function (see above)

File: backend/test/cache.spec.js [hunk 14]
```diff
@@ -79,6 +79,13 @@ describe('cache', function () {
     expect(stub).toBeCalledTimes(1)
   })
 
+  it('should dispatch "getResourceQuotas" to internal cache', function () {
+    const list = []
+    const stub = jest.spyOn(internalCache, 'getResourceQuotas').mockReturnValue(list)
+    expect(cache.getResourceQuotas()).toEqual(list)
```
Comment (holgerkoser) [inline backend/test/cache.spec.js:85]:
```suggestion
    expect(cache.getResourceQuotas()).toBe(list)
```

File: frontend/src/views/Administration.vue [hunk 15]
```diff
@@ -547,6 +627,16 @@ export default {
       this.detailedMessage = undefined
       this.edit = false
     }
+  },
+  created () {
+    // see https://router.vuejs.org/guide/advanced/data-fetching.html#fetching-after-navigation
+    this.$watch(
+      () => this.$route.params,
+      () => {
+        this.fetchProjectQuota(this.project.metadata.namespace)
```
Comment (holgerkoser) [inline frontend/src/views/Administration.vue:636]:
```suggestion
        try {
           this.fetchProjectQuota(this.project.metadata.namespace)
        } catch (err) {
          this.$logger.error('Failed to fetch project quota: %s', err.message)
          // or maybe snotify
        } 
```

File: frontend/src/utils/index.js [hunk 16]
```diff
@@ -317,6 +320,57 @@ export function getProjectDetails (project) {
...
+    const resourceName = replace(key, 'count/', '')
+    const caption = upperFirst(head(split(resourceName, '.')))
+    let progressColor = 'primary'
+
+    if (percentage >= 80) {
+      progressColor = 'warning'
+    }
+    if (percentage >= 100) {
+      progressColor = 'error'
+    }
```
Comment (holgerkoser) [inline frontend/src/utils/index.js:341]:
```suggestion
    if (percentage >= 100) {
      progressColor = 'error'
    } else if (percentage >= 80) {
      progressColor = 'warning'
    }

```

File: frontend/src/utils/index.js [hunk 17]
```diff
@@ -317,6 +320,57 @@ export function getProjectDetails (project) {
   }
 }
 
+export function getProjectQuotaStatus (projectQuota) {
```
Comment (holgerkoser) [inline frontend/src/utils/index.js:323]: I would move these two functions into the store. In the tests we can then import the raw getters.

File: backend/lib/services/index.js (hunk 12 above)
Comment (grolu) [reply backend/lib/services/index.js:19]: changed also for cloudprovidersecrets

File: frontend/tests/unit/store.projectQuota.spec.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
+//
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import { getProjectQuotaStatus, aggregateResourceQuotaStatus } from '@/store/modules/projectQuota'
```
Comment (holgerkoser) [inline frontend/tests/unit/store.projectQuota.spec.js:7]: I would not export these two helper function from the store and instead would add the aggregation to the mutation and call it `SET_RESOURCE_QUOTAS`:

```js
// actions
const actions = {
  async fetchProjectQuota ({ commit }, namespace) {
    const { data } = await getResourceQuotas({ namespace })
    commit('SET_RESOURCE_QUOTAS', [namespace, data])
  }
}

// mutations
const mutations = {
  SET_RESOURCE_QUOTAS (state, [key, quotas]) {
    const value = aggregateResourceQuotaStatus(map(quotas, 'status'))
    Vue.set(state, key, value)
  }
}
```

The tests can remain the same if we defined the two function locally by calling the getter and mutation:
```js
import projectQuota from '@/store/modules/projectQuota'

function getProjectQuotaStatus (quota) {
  const namespace = 'default'
  return projectQuota.getters.status({ [namespace]: quota }, null, { namespace })
}

function aggregateResourceQuotaStatus (statuses) {
  const namespace = 'default'
  const state = {}
  const quotas = statuses.map(status => ({ status }))
  projectQuota.mutations.SET_RESOURCE_QUOTAS(state, [namespace, quotas])
  return state[namespace]
}
```

Comment (grolu) [reply frontend/tests/unit/store.projectQuota.spec.js]: ok.

**Review by petersutter — APPROVED** [review]
/lgtm

**Pull request merged**